File: src/localization_pkg/src/livox_custom_to_pc2_node.cpp
#include <cmath>
#include <cstdint>
#include <cstring>
#include <string>
#include <vector>

#include "livox_ros_driver2/msg/custom_msg.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/msg/point_field.hpp"

namespace
{
constexpr uint32_t kPointStep = 16;  // x, y, z, intensity as float32

void write_float(std::vector<uint8_t> & data, const size_t offset, const float value)
{
  std::memcpy(data.data() + offset, &value, sizeof(float));
}
}  // namespace

class LivoxCustomToPc2Node : public rclcpp::Node
{
public:
  LivoxCustomToPc2Node()
  : Node("livox_custom_to_pc2")
  {
    auto qos = rclcpp::SensorDataQoS();
    sub_ = create_subscription<livox_ros_driver2::msg::CustomMsg>(
      "/livox/lidar", qos,
      std::bind(&LivoxCustomToPc2Node::callback, this, std::placeholders::_1));
    pub_ = create_publisher<sensor_msgs::msg::PointCloud2>("/livox/lidar_pc2", qos);

    RCLCPP_INFO(
      get_logger(),
      "livox_custom_to_pc2: C++ relay ready  /livox/lidar -> /livox/lidar_pc2");
  }

private:
  void callback(const livox_ros_driver2::msg::CustomMsg::SharedPtr msg)
  {
    if (msg->points.empty()) {
      return;
    }

    sensor_msgs::msg::PointCloud2 out;
    out.header = msg->header;
    out.height = 1;
    out.is_bigendian = false;
    out.is_dense = true;
    out.point_step = kPointStep;
    out.fields.resize(4);

    out.fields[0].name = "x";
    out.fields[0].offset = 0;
    out.fields[0].datatype = sensor_msgs::msg::PointField::FLOAT32;
    out.fields[0].count = 1;

    out.fields[1].name = "y";
    out.fields[1].offset = 4;
    out.fields[1].datatype = sensor_msgs::msg::PointField::FLOAT32;
    out.fields[1].count = 1;

    out.fields[2].name = "z";
    out.fields[2].offset = 8;
    out.fields[2].datatype = sensor_msgs::msg::PointField::FLOAT32;
    out.fields[2].count = 1;

    out.fields[3].name = "intensity";
    out.fields[3].offset = 12;
    out.fields[3].datatype = sensor_msgs::msg::PointField::FLOAT32;
    out.fields[3].count = 1;

    out.data.reserve(msg->points.size() * kPointStep);

    for (const auto & point : msg->points) {
      if (!std::isfinite(point.x) || !std::isfinite(point.y) || !std::isfinite(point.z)) {
        continue;
      }

      const size_t offset = out.data.size();
      out.data.resize(offset + kPointStep);
      write_float(out.data, offset + 0, point.x);
      write_float(out.data, offset + 4, point.y);
      write_float(out.data, offset + 8, point.z);
      write_float(out.data, offset + 12, static_cast<float>(point.reflectivity));
    }

    out.width = static_cast<uint32_t>(out.data.size() / kPointStep);
    if (out.width == 0) {
      return;
    }
    out.row_step = out.point_step * out.width;
    pub_->publish(out);
  }

  rclcpp::Subscription<livox_ros_driver2::msg::CustomMsg>::SharedPtr sub_;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr pub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<LivoxCustomToPc2Node>());
  rclcpp::shutdown();
  return 0;
}

File: src/localization_pkg/scripts/livox_custom_to_pc2_node.py
#!/usr/bin/env python3
"""
livox_custom_to_pc2_node.py — Relay: Livox CustomMsg → PointCloud2.

Required when xfer_format=1 (fast_lio_mode): FAST-LIO needs CustomMsg on
/livox/lidar, but GICP and pointcloud_to_laserscan need PointCloud2.

Topics:
  Input:  /livox/lidar      livox_ros_driver2/msg/CustomMsg
  Output: /livox/lidar_pc2  sensor_msgs/PointCloud2  (fields: x, y, z, intensity)
"""
import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, HistoryPolicy, QoSProfile, ReliabilityPolicy
from livox_ros_driver2.msg import CustomMsg
from sensor_msgs.msg import PointCloud2, PointField


_FIELDS_XYZI = [
    PointField(name='x',         offset=0,  datatype=PointField.FLOAT32, count=1),
    PointField(name='y',         offset=4,  datatype=PointField.FLOAT32, count=1),
    PointField(name='z',         offset=8,  datatype=PointField.FLOAT32, count=1),
    PointField(name='intensity', offset=12, datatype=PointField.FLOAT32, count=1),
]
_POINT_STEP = 16  # 4 × float32


class LivoxCustomToPc2Node(Node):
    def __init__(self):
        super().__init__('livox_custom_to_pc2')

        sensor_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=5,
        )

        self._sub = self.create_subscription(
            CustomMsg, '/livox/lidar', self._cb, sensor_qos
        )
        self._pub = self.create_publisher(
            PointCloud2, '/livox/lidar_pc2', sensor_qos
        )
        self.get_logger().info('livox_custom_to_pc2: relay ready  /livox/lidar → /livox/lidar_pc2')

    def _cb(self, msg: CustomMsg) -> None:
        points = msg.points
        n = len(points)
        if n == 0:
            return

        # Build Nx4 float32 array in one pass using list comprehension.
        # For 20 k points at 10 Hz this takes ~3 ms in CPython — acceptable.
        raw = np.array(
            [[p.x, p.y, p.z, float(p.reflectivity)] for p in points],
            dtype=np.float32,
        )

        # Drop points with non-finite coordinates (origin zeros from missed returns)
        valid = np.isfinite(raw[:, 0]) & np.isfinite(raw[:, 1]) & np.isfinite(raw[:, 2])
        raw = np.ascontiguousarray(raw[valid])
        n = raw.shape[0]
        if n == 0:
            return

        out = PointCloud2()
        out.header = msg.header
        out.height = 1
        out.width = n
        out.fields = _FIELDS_XYZI
        out.is_bigendian = False
        out.point_step = _POINT_STEP
        out.row_step = _POINT_STEP * n
        out.is_dense = True
        out.data = raw.tobytes()
        self._pub.publish(out)


def main(args=None):
    rclpy.init(args=args)
    node = LivoxCustomToPc2Node()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
